File: cpp/1026.maximum-difference-between-node-and-ancestor.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxAncestorDiff(TreeNode* root) {
        if (root == NULL) {
            return 0;
        }

        stack<tuple<TreeNode*, int, int>> st;
        st.push({ root, root->val, root->val });
        int res = 0;

        while (!st.empty()) {
            auto [node, lo, hi] = st.top(); st.pop();

            if (node == NULL) {
                res = max(res, hi - lo);
                continue;
            }

            lo = min(lo, node->val);
            hi = max(hi, node->val);

            st.push({node->right, lo, hi});
            st.push({node->left, lo, hi});
        }

        return res;
    }
};

File: golang/1026.maximum-difference-between-node-and-ancestor.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

type State struct {
    node *TreeNode
    lo, hi int
}

func maxAncestorDiff(root *TreeNode) int {
    if root == nil {
        return 0
    }

    res := 0

    st := []State{ State{ root, root.Val, root.Val } }
    var top State
    var lo, hi int

    for len(st) > 0 {
        top = st[len(st) - 1]
        st = st[:len(st) - 1]

        if top.node == nil {
            res = max(res, top.hi - top.lo)
            continue
        }

        lo = min(top.lo, top.node.Val)
        hi = max(top.hi, top.node.Val)

        st = append(st, State{ top.node.Right, lo, hi })
        st = append(st, State{ top.node.Left, lo, hi })
    }

    return res
}

func min(a, b int) int {
    if a < b {
        return a
    }
    return b
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}
